src/platform/android/java/chip/platform/AndroidChipPlatform.java

```diff
@@ -0,0 +1,80 @@
+/*
+ *   Copyright (c) 2020 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package chip.platform;
+
+public final class AndroidChipPlatform {
+  private static final String TAG = AndroidChipPlatform.class.getSimpleName();
+  private static AndroidChipPlatform sInstance = new AndroidChipPlatform();
+  private BLEManager mBLEManager = null;
+
+  private AndroidChipPlatform() {}
+
+  public static AndroidChipPlatform getInstance() {
+    return sInstance;
+  }
+
+  // for BLEManager
+  public BLEManager getBLEManager() {
+    return mBLEManager;
+  }
+
+  public void setBLEManager(BLEManager manager) {
+    if (mBLEManager == null) {
+      mBLEManager = manager;
+      nativeSetBLEManager(manager);
+    }
+  }
+
+  public native void nativeSetBLEManager(BLEManager manager);
+
+  // apis in BleLayer.h called by Platform
+  // write success
+  public native void handleWriteConfirmation(
+      int connId, byte[] svcId, byte[] charId, boolean success);
+
+  // onSubscribeCharacteristic get data
+  public native void handleIndicationReceived(int connId, byte[] svcId, byte[] charId, byte[] data);
+
+  // Subscribe success
+  public native void handleSubscribeComplete(
+      int connId, byte[] svcId, byte[] charId, boolean success);
+
+  // Unsubscribe success
+  public native void handleUnsubscribeComplete(
+      int connId, byte[] svcId, byte[] charId, boolean success);
+
+  // connection status changed
+  public native void handleConnectionError(int connId);
+
+  // for KeyValueStoreManager
+  public native void setKeyValueStoreManager(KeyValueStoreManager manager);
+
+  // for ConfigurationManager
+  public native void setConfigurationManager(ConfigurationManager manager);
+
+  // for ServiceResolver
+  public native void setServiceResolver(ServiceResolver resolver);
+
+  public native void handleServiceResolve(
```

Comment:
It looks like you overwrote the changes from #9097. Could we move back to having a callback interface for this?

Reply:
Yes, I noticed this change, is that to avoid singleton here? Now I have avoid singleton in AndroidChipPlatform and the object which hold handleServiceResolve are also setting to NsdManagerServiceResolver but by app. Do you think it is OK?

Reply:
Yes, that's fine. It's actually a more flexible design, because it allows the app to pick what Android mDNS implementation they want to use.

In general, that's why I want to avoid singletons except at the app level: allow for components to be flexibly swapped out.

Reply:
If so, I think it works now. In fact, the dns-sd in android framework do not supports subtype, we are plaining to implement a new one in our app.

Reply:
I didn't see any changes here. To be clear, I want to restore the changes from #9097:
1. ChipMdnsCallback should be a constructor parameter of AndroidChipPlatform
2. ServiceResolver should not have a hard-coded `setAndroidChipPlatform` method, but instead pass `ChipMdnsCallback` as a parameter to `resolve` (see above PR).

Reply:
My understanding, as #9097, the app can have their own UserServiceResolver to rewrite resolver, and also their own UserChipMdnsCallbackImpl to do something before calling ChipMdnsCallbackImpl::HandleResolve, but at last, he has no choice but calling it.
![Matter_SDK-NDSAndroid drawio (1)](https://user-images.githubusercontent.com/1230620/135038833-8d09b498-75cd-4746-9e7f-e2b5a6a767f5.png)

While in current design, the app can also have their own UserServiceResolver to rewrite resolver, and also when he get the resolver result, he can also do something before calling AndroidChipPlatform::HandleResolve, also he has no choice but calling it. 
![Matter_SDK-NDSAndroid drawio (2)](https://user-images.githubusercontent.com/1230620/135039199-2a8cb4c5-0364-45ba-8b66-1b45acffda1e.png)

No more can do and no less can do, but it is more easy because we do not need to pass a ChipMdnsCallback from java to native and native to java. Anything I am missed?

Reply:
I understand the platform `HandleResolve` has to be called no matter what. My rationale for using ChipMdnsCallback/ChipMdnsCallback is more about flexibility for testing.

Let's say you have a Robolectric unit test for your app that happens to use `ServiceResolver` and `handleServiceResolve`. Since this is a Robolectric unit test, you are not going to use the real dnssd, and you would need some complicated setup to use native methods.

For `ServiceResolver` this is fine, because you can swap out the real `NsdManagerServiceResolver` for a dummy `FakeServiceResolver` that doesn't actually try to discover things on the network.

For `AndroidChipPlatform.handleServiceResolve`, you are stuck with the real implementation in your unit test, which may be problematic since it calls native methods. Whereas, if there is a `ChipMdnsCallback` interface, you can swap the real `ChipMdnsCallback` with some `FakeChipMdnsCallback` that doesn't use native methods.

I would not have asked you to implement this as a new feature. I'm just asking to keep the existing functionality.

Reply:
I understand now, fixed at ba37f57